cnb_index.go

```diff
@@ -35,161 +35,144 @@ type CNBIndex struct {
 	RepoName string
 }
 
-func (h *CNBIndex) getConfigFileFrom(digest name.Digest) (v1.ConfigFile, error) {
-	hash, err := v1.NewHash(digest.Identifier())
-	if err != nil {
-		return v1.ConfigFile{}, err
-	}
-	image, err := h.Image(hash)
-	if err != nil {
-		return v1.ConfigFile{}, err
-	}
-	configFile, err := GetConfigFile(image)
-	if err != nil {
-		return v1.ConfigFile{}, err
-	}
-	return *configFile, nil
-}
-
-func (h *CNBIndex) getManifestFileFrom(digest name.Digest) (v1.Manifest, error) {
-	hash, err := v1.NewHash(digest.Identifier())
-	if err != nil {
-		return v1.Manifest{}, err
-	}
-	image, err := h.Image(hash)
+func (h *CNBIndex) getDescriptorFrom(digest name.Digest) (v1.Descriptor, error) {
+	indexManifest, err := getIndexManifest(h.ImageIndex)
 	if err != nil {
-		return v1.Manifest{}, err
+		return v1.Descriptor{}, err
 	}
-	manifestFile, err := GetManifest(image)
-	if err != nil {
-		return v1.Manifest{}, err
+	for _, current := range indexManifest.Manifests {
+		if current.Digest.String() == digest.Identifier() {
+			return current, nil
+		}
 	}
-	return *manifestFile, nil
+	return v1.Descriptor{}, fmt.Errorf("failed to find image with digest %s in index", digest.Identifier())
 }
 
 // OS returns `OS` of an existing Image.
 func (h *CNBIndex) OS(digest name.Digest) (os string, err error) {
-	configFile, err := h.getConfigFileFrom(digest)
+	desc, err := h.getDescriptorFrom(digest)
 	if err != nil {
 		return "", err
 	}
-	return configFile.OS, nil
+	if desc.Platform != nil {
```

Comment:
nit: `getDescriptorFrom` never returns nil in its first argument, so I think you can remove this check

Reply:
I think I added it because on unit testing when using `random` packages in `ggcr` some nil in these attributes but let me check if everything is file after removing it